data/account/participationRegistry.go

```diff
@@ -0,0 +1,393 @@
+// Copyright (C) 2019-2021 Algorand, Inc.
+// This file is part of go-algorand
+//
+// go-algorand is free software: you can redistribute it and/or modify
+// it under the terms of the GNU Affero General Public License as
+// published by the Free Software Foundation, either version 3 of the
+// License, or (at your option) any later version.
+//
+// go-algorand is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU Affero General Public License for more details.
+//
+// You should have received a copy of the GNU Affero General Public License
+// along with go-algorand.  If not, see <https://www.gnu.org/licenses/>.
+
+package account
+
+import (
+	"context"
+	"database/sql"
+	"errors"
+	"fmt"
+
+	"github.com/algorand/go-algorand/crypto"
+	"github.com/algorand/go-algorand/data/basics"
+	"github.com/algorand/go-algorand/util/db"
+)
+
+const maxBalLookback = 320
```

Comment:
@tsachiherman I don't want to bring in a dependency on the consensus params here, so I'd like to remove this and simply record the registration round:
`EffectiveFirst` -> `RegisteredAt`
`EffectiveLast` -> `DeregisteredAt`

Also, I think I have an off-by-one error and `EffectiveLast` == `EffectiveFirst` when replacing a key. Do you happen to know if `EffectiveLast` should be -1 or `EffectiveFirst` should be +1?

Reply:
The `RegisteredAt` is very different from `EffectiveFirst`, and `EffectiveLast` is very different from `DeregisteredAt`. While I don't mind storing these, there is no "notable" gains here.
Btw - I always used the term "register/unregister".. so "deregister" sounds a bit weird. I went looking, and found this - https://english.stackexchange.com/questions/25931/unregister-vs-deregister. so, it looks like you're correct : since you're denoting the action rather then the state, it should be `Deregister` and it would make it's state `unregistered`.

As for the other question, the effective round should be real round numbers where the node can use the voting key. so, having these equal is legit.

Reply:
For the second question let me clarify. When I'm overriding key1 with key2, I'm not sure which round to use for key1.EffectiveLast and key2.EffectiveFirst.

Currently I have it so `key1.EffectiveLast = key2.EffectiveFirst = currentRound + 320`, I'm confident that at least one of those is wrong.